### DISCUSSION TITLE
Delegated requests for embedded language don't match results for when language is not embedded

### DISCUSSION THREAD

**Discussion by vinistock** [microsoft/vscode-discussions#1628] [Extension Development QnA] [answered]
- VS Code Version: 1.94.0-insider Commit: c4efe1dc9eec4914f3076b2d954fe4fe174a5820

### Context

I'm working on adding [request delegation for ERB files in the Ruby language server](https://github.com/Shopify/ruby-lsp/pull/2551). I followed the instructions in the [request forwarding sample](https://github.com/microsoft/vscode-extension-samples/tree/main/lsp-embedded-request-forwarding). I'm seeing some weird behaviour, which could 100% be a problem in my understanding or a bug in VS Code.

ERB stands for embedded Ruby and it's a way to use Ruby code to produce a result in a host language (commonly used with HTML). In the PR, I'm using exactly the same technique as the sample to create a virtual document with only the HTML content and every embedded Ruby part stripped out.

Then I delegate the completion request using the VS Code built-in command, again exactly like the sample:

```typescript
// Code used to delegate a completion request inside the HTML part of an ERB file to the
// appropriate language service

vscode.commands.executeCommand<CompletionList>(
  "vscode.executeCompletionItemProvider",
  vscode.Uri.parse(vdocUriString),
  params.position,
  params.context.triggerCharacter,
)
```

### Expected vs actual results

The request delegation is working properly for some cases. For example, I get the right completion for HTML tags and attributes.

The thing that called my attention is that the completion results are not the same for the delegated request vs requests to an actual HTML file, even if the contents match exactly.

For example, if I create an ERB template that doesn't include any Ruby, just pure HTML, the delegate completion results don't match what I get for the exact same source if the language mode is HTML (and thus no delegation is involved).

```erb
<!--- There's no Ruby in this ERB template, so the virtual HTML content is exactly the same -->
<script type="text/javascript">
  document.
</script>
```

The completion for the delegate request seems to return generic completion items, without considering the `document` as the receiver. For example, `getElementById` is not available as part of the results.

```typescript
['$', 'AbortController', 'AbortSignal', 'AbstractRange', 'ActiveXObject', 'AnalyserNode', 'Animation', 'AnimationEffect', 'AnimationEvent', 'AnimationPlaybackEvent', 'AnimationTimeline', 'Array', 'ArrayBuffer', ...]
```

Whereas when the language mode is HTML and no delegation is involved, I get accurate completion that includes the expected functions. I expected the results for the delegated request to be the exact same.

### Question

I know that the language server bundled in the built-in `html-language-features` extension also performs request delegation to get the right features for embedded JavaScript and CSS in the middle of HTML.

Is the problem that the request is being delegated twice? The funny thing is that I do get CSS and JavaScript completion when inside the right contexts, it's just the JavaScript completion that seems to be missing context. Almost as if the code is not being type checked before gathering completions.

Does the extension need to delegate something else before starting to request completions?

I would really appreciate any pointers as to why this is happening.

**Comment by isidorn** [comment]
@mjbvz could you help please?

**Reply by mjbvz** [reply]
@aeschli Can you please take a look since this is related to the built-in html extension

**Reply by aeschli** [reply, marked as answer]
This should work. I wonder if the position is really after the `.`.  
To investigate, can you modify the [request forwarding sample](https://github.com/microsoft/vscode-extension-samples/tree/main/lsp-embedded-request-forwarding) in such a way that it reproduces the problem?

**Reply by burkeholland** [reply]
@vinistock Can you verify if this fixes the issue?

**Reply by vinistock** [reply]
Sorry, I haven't had the time to create a reproduction yet. I'll try to get one going.

**Reply by burkeholland** [reply]
Closing this for now under lack of activity. If you feel this is a mistake, please feel free to re-open. Thank you so much!

**Reply by connorshea** [reply]
I looked into this, it looks like the embedded-content URI is not parsed quite correctly between `vscode.Uri.parse` and the encoding recommended as part of the documentation for forwarding [embedded content](https://code.visualstudio.com/api/language-extensions/embedded-languages#request-forwarding).

Very rudimentary example, but basically the exact same thing that's happening to me:
```ts
import { URI as Uri } from 'vscode-uri';

const originalUri = "file:///Users/connorshea/code/vglist/app/views/engines/new.html.erb";
console.log(`originalUri: ${originalUri}`);
const vdocUriString = `embedded-content://html/${encodeURIComponent(originalUri)}.html`;
console.log(`vdocUriString: ${vdocUriString}`);
const vdocUri = Uri.parse(vdocUriString);
console.log(`vdocUri: ${vdocUri}`);
```

output:

```
originalUri: file:///Users/connorshea/code/vglist/app/views/engines/new.html.erb
vdocUriString: embedded-content://html/file%3A%2F%2F%2FUsers%2Fconnorshea%2Fcode%2Fvglist%2Fapp%2Fviews%2Fengines%2Fnew.html.erb.html
vdocUri: embedded-content://html/file%3A///Users/connorshea/code/vglist/app/views/engines/new.html.erb.html
```

Note in the `vdocUri` that the colon in `file:///` is still encoded. This is basically copy-pasted from the Request Forwarding Sample in the API docs (albeit with `css` swapped for `html`): https://code.visualstudio.com/api/language-extensions/embedded-languages#request-forwarding-sample

It is also basically identical to the behavior in the code here if you gave it a `file://` url:
https://github.com/microsoft/vscode-extension-samples/blob/32c5dda0d58a9d92952498e8934e574f1e099ff0/lsp-embedded-request-forwarding/client/src/extension.ts#L49-L55

And is what's going on in the Ruby LSP code:

https://github.com/Shopify/ruby-lsp/blob/e23c1a77cc7cf5e84c16c42e4a893f8ae4e9c07d/vscode/src/client.ts#L573-L586

Is there a good way for the Ruby LSP folks to fix this? It might be a bug in the vscode-uri parser that only shows up for the `embedded-content://blah/file:///Users/foobar/etc/etc/baz.html.example`-style of URL?

Basically, the current implementation results in the URL passed forward from the Ruby LSP into the HTML Language Service and then forward onto TypeScript being "corrupted" in the exact same way (and showing the same problem as the video noted above in the original post):

<img width="1092" height="513" alt="Screenshot 2025-10-12 at 10 54 08 PM" src="https://github.com/user-attachments/assets/862526c6-dee7-4bf9-892e-0360f0e74eab" />

I tried messing with how the Ruby LSP encodes this URL, but I couldn't figure out a way to get it to properly encode/decode at all levels without breaking _something_. And encoding it this way _is_ literally what the VS Code docs/examples say to do.

**Reply by vinistock** [reply]
@connorshea thank you so much for diving deep into this, I didn't have time to study it in this detail.

Folks, please let us know if this needs to be fixed in VS Code or if there's some workaround we can apply on our extension 🙏.
